Implement native support StringViewArray for `regexp_is_match` and `regexp_is_match_scalar` function, deprecate `regexp_is_match_utf8` and `regexp_is_match_utf8_scalar`

tlm365 opened the pull request:
# Which issue does this PR close?
Closes #6370.
<!--
We generally require a GitHub issue to be filed for all bug fixes and enhancements and this helps us generate change logs for our releases. You can link an issue to this PR using the GitHub syntax. For example `Closes #123` indicates that this PR will close issue #123.
-->

# Rationale for this change
1. Natively operate on StringViewArray without having to convert first to StringArray
2. (Potentially) take advantage of the new string view layout

<!--
Why are you proposing this change? If this is already explained clearly in the issue then this section is not needed.
Explaining clearly why changes are proposed helps reviewers understand your changes and offer better suggestions for fixes.
-->

# What changes are included in this PR?
Introduce `regexp_is_match` and `regexp_is_match_scalar` (which can replace `regexp_is_match_utf8` and `regexp_is_match_utf8_scalar`) can perform on `StringArray` / `LargeStringArray` / `StringViewArray` arguments.
<!--
There is no need to duplicate the description in the issue here but it is sometimes worth providing a summary of the individual changes in this PR.
-->

# Are there any user-facing changes?
No.
<!--
If there are user-facing changes then we may require documentation to be updated before approving the PR.
-->

<!---
If there are any breaking changes to public APIs, please add the `breaking change` label.
-->

alamb commented on arrow-string/src/regexp.rs:
```diff
@@ -591,40 +604,103 @@ mod tests {
+    );
+    test_flag_utf8!(
+        test_utf8_array_regexp_is_match_insensitive_2,
+        StringViewArray::from(vec!["arrow", "arrow", "arrow", "arrow", "arrow", "arrow"]),
```
StringViewArray has special case handling for strings that are more than 12 bytes long (the string data is stored out of band in those cases)

Can you please add tests that have some strings that are longer than 12 bytes?

tlm365 replied:
@alamb Thanks for reviewing. I'm willing to add benchmark for this one.

> Can you please add tests that have some strings that are longer than 12 bytes?

Yes, noted. I will review and update test cases for this scenario.

alamb commented on arrow-string/src/regexp.rs:
```diff
@@ -107,18 +116,22 @@ pub fn regexp_is_match_utf8<OffsetSize: OffsetSizeTrait>(
 /// See the documentation on [`regexp_is_match_utf8`] for more details.
-pub fn regexp_is_match_utf8_scalar<OffsetSize: OffsetSizeTrait>(
-    array: &GenericStringArray<OffsetSize>,
+pub fn regexp_is_match_utf8_scalar<'a, S>(
```
Unfortunately, I think this is a API change (as is the above)

alamb replied:
I have an idea of how to update this PR to avoid an API change -- the reason this is important is that a breaking API change would need to wait until the next major release (Dec 2024) per the release schedule: https://github.com/apache/arrow-rs?tab=readme-ov-file#release-versioning-and-schedule

TLDR is I think if we introduced a new function like the following: 

```rust
fn regexp_is_match(
    array: &dyn Array, 
    regex_array: &dyn Array, 
    flags_array: Option<&dyn Array, >,
) -> Result<BooleanArray, ArrowError> {
..
}
``

We could then support StringView and StringArray and LargeStringArray

tlm365 replied:
> TLDR is I think if we introduced a new function like the following:

Sounds good 👍 But why do we use `&dyn Array` for the new `regex_is_match` function instead of keeping the current implementation?

Or am I misunderstanding you? I understand that we will provide a new `regex_is_match` function, and mark the current `regex_is_match_utf8` function as:
```rust
#[deprecated(since="54.0.0", note="please use `regex_is_match` instead")]
pub fn regexp_is_match_utf8(...) { ... }
```
Is that right? 🤔

alamb reviewed (commented):
Thank you @tlm365  -- this is looking really nice

I wonder if you might also be willing to add StringView to the benchmarks as well, specifically

https://github.com/apache/arrow-rs/blob/704f90bbf541896387bed030e12a39be308047e8/arrow/benches/comparison_kernels.rs#L56-L62

So that if this code is changed in the future we can ensure it doesn't regress in performance

tlm365 commented:
> I wonder if you might also be willing to add StringView to the benchmarks as well, specifically
> So that if this code is changed in the future we can ensure it doesn't regress in performance

@alamb Thanks for reviewing, willing to add benchmark for this one. I will update it soon.

alamb commented:
Here are the benchmark results (aka this PR doesn't slow down the existing implementation)

```
++ critcmp master regex-is-match-utf8
group                                                     master                                 regex-is-match-utf8
-----                                                     ------                                 -------------------
regexp_matches_utf8 scalar ends with                      1.02  1932.3±20.07µs        ? ?/sec    1.00  1898.4±17.40µs        ? ?/sec
regexp_matches_utf8 scalar starts with                    1.00  1932.1±14.07µs        ? ?/sec    1.00  1924.9±26.17µs        ? ?/sec
```

alamb commented:
I am depressed about the large review backlog in this crate. We are looking for more help from the community reviewing PRs -- see https://github.com/apache/arrow-rs/issues/6418 for more

alamb commented on arrow-string/src/regexp.rs:
```diff
@@ -35,16 +38,64 @@ use std::sync::Arc;
+/// * [`regexp_is_match_scalar`] for matching a single regular expression against an array of strings
+/// * [`regexp_match`] for extracting groups from a string array based on a regular expression
+///
+/// # Example
```
I added an example to help the migration

![Screenshot 2024-09-18 at 5 40 38 PM](https://github.com/user-attachments/assets/05ef31e0-3d1b-4f19-8291-cb0b0800ef5e)

alamb commented on arrow-string/src/regexp.rs:
```diff
@@ -35,16 +38,64 @@ use std::sync::Arc;
     regex_array: &GenericStringArray<OffsetSize>,
     flags_array: Option<&GenericStringArray<OffsetSize>>,
 ) -> Result<BooleanArray, ArrowError> {
+    regexp_is_match(array, regex_array, flags_array)
```
I switched the implementation to just call the new function to avoid duplication

alamb reviewed (approved):
Thank you very much @tlm365  -- this looks great.

I was reviewing this PR and I had the code checked out locally, so I took the liberty of making a few changes:

1. I fixed clippy (was failing due to using deprecated functions)
1. I updated the comments / added an example to ease the transition (passing `None` as the flags argument results in type inference errors without some type help)
2. Improved the comments in general making it clearer what regexp_is_match does and how it is related to `regexp_match`

alamb commented:
@tlm365  I wonder if you have a few minutes to review the changes I pushed to this PR. 

I again I am sorry about the review delays

tlm365 commented:
> @tlm365 I wonder if you have a few minutes to review the changes I pushed to this PR.
> 
> I again I am sorry about the review delays

@alamb Oops, thank you so much for reviewing. Sorry 🙇 I've been a little busy lately. Noted and will come back to review this weekend.

tlm365 commented:
> Thank you very much @tlm365 -- this looks great.
> 
> I was reviewing this PR and I had the code checked out locally, so I took the liberty of making a few changes:
> 
> 1. I fixed clippy (was failing due to using deprecated functions)
> 2. I updated the comments / added an example to ease the transition (passing `None` as the flags argument results in type inference errors without some type help)
> 3. Improved the comments in general making it clearer what regexp_is_match does and how it is related to `regexp_match`

@alamb it looks very nice 👍 thank you so much for this update! ❤️

Dandandan commented:
Thanks @tlm365 and @alamb

alamb commented:
We updated this PR so it was not an API change so removing the label
